### PR TITLE
[network-data] always support service registration at Leader

### PR DESCRIPTION
**Pull request by librasungirl**
The thought behind this commit is that Leader should be able to
process service registration by default, like prefix registration.

This commit:
- limit SERVICE to wrap code relative to service provider.
- support service traffic forwarding when SERVICE is disabled (as
  a following up update after #4219).
- other minor updates.

### CODE REVIEW

File: src/core/thread/network_data_local.cpp [hunk 1]
```diff
@@ -175,6 +176,53 @@ otError Local::RemoveHasRoutePrefix(const uint8_t *aPrefix, uint8_t aPrefixLengt
...
+            ContainsOnMeshPrefixes(Get<Leader>(), Get<Mle::MleRouter>().GetRloc16()));
+}
+
+bool Local::IsExternalRouteConsistent(void)
+{
+    return (Get<Leader>().ContainsExternalRoutes(*this, Get<Mle::MleRouter>().GetRloc16()) &&
+            ContainsExternalRoutes(Get<Leader>(), Get<Mle::MleRouter>().GetRloc16()));
+}
+
+#endif
```
Comment (jwhui) [inline src/core/thread/network_data_local.cpp]: nit: add `// OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE`

File: src/core/thread/network_data_local.cpp [hunk 2]
```diff
@@ -240,111 +288,65 @@ otError Local::RemoveService(uint32_t aEnterpriseNumber, const uint8_t *aService
...
+    return (Get<Leader>().ContainsServices(*this, Get<Mle::MleRouter>().GetRloc16()) &&
+            ContainsServices(Get<Leader>(), Get<Mle::MleRouter>().GetRloc16()));
 }
 
-void Local::UpdateRloc(BorderRouterTlv &aBorderRouter)
-{
-    BorderRouterEntry *entry = aBorderRouter.GetEntry(0);
-    entry->SetRloc(Get<Mle::MleRouter>().GetRloc16());
-}
+#endif
```
Comment (jwhui) [inline src/core/thread/network_data_local.cpp]: nit: add `// OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE`

File: src/core/thread/network_data_local.cpp (hunk 1 above)
Comment (librasungirl) [reply src/core/thread/network_data_local.cpp]: updated

File: src/core/thread/network_data_local.cpp (hunk 2 above)
Comment (librasungirl) [reply src/core/thread/network_data_local.cpp]: updated.

File: src/core/thread/network_data_local.cpp [hunk 3]
```diff
@@ -41,17 +41,18 @@
 #include "mac/mac_types.hpp"
 #include "thread/thread_netif.hpp"
 
-#if OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE || OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
-
 namespace ot {
 namespace NetworkData {
 
+#if OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE || OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
```
Comment (jwhui) [inline src/core/thread/network_data_local.cpp]: Any reason to move this inside the namespace?

File: src/core/thread/network_data_local.hpp [hunk 4]
```diff
@@ -51,6 +51,7 @@ namespace ot {
 
 namespace NetworkData {
 
+#if OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE || OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
```
Comment (jwhui) [inline src/core/thread/network_data_local.hpp]: Move outside namespace?

**Pull request merged**